File: C++/MergeSort.cpp
#include "MergeSort.h"
#include <iostream>
#include <fstream>
#include <sstream>
#include <string>
#include <vector>
#include <chrono>
#include <cstring>


int main(int argc, char** args)
{
	std::string inputFileName = args[1];
	std::string outputFileName = args[2];
	int iterations = std::stoi(args[3]);


	std::vector<int> lineValues = std::vector<int>();
	long long* executionsTimes = new long long[iterations];

	{
		std::ifstream inFile = std::ifstream(inputFileName);
		int num;
		while (inFile >> num)
		{
			lineValues.push_back(num);
		}
		inFile.close();
	}


	size_t dataSize = lineValues.size();
	int* data = new int[dataSize];
	int* dataCopy = new int[dataSize];
	for (size_t i = 0; i < dataSize; i++)
	{
		data[i] = lineValues.at(i);
	}

	for (size_t i = 0; i < iterations; i++)
	{
		std::memcpy(dataCopy, data, sizeof(int) * dataSize);
		auto start = std::chrono::high_resolution_clock::now();
		mergeSort(data, 0, dataSize - 1);
		auto end = std::chrono::high_resolution_clock::now();
		std::chrono::nanoseconds time = std::chrono::duration_cast<std::chrono::nanoseconds>(end - start);
		executionsTimes[i] = time.count();
	}

	{
		std::ofstream outFile;
		outFile.open(outputFileName);

		for (size_t i = 0; i < iterations; i++)
		{
			outFile << executionsTimes[i] << std::endl;
		}

		outFile.close();
	}

	delete[](data);
}

void mergeSort(int* data, int l, int r)
{
	if (l >= r)
	{
		return;
	}
	int m = l + (r - l) / 2;
	mergeSort(data, l, m);
	mergeSort(data, m + 1, r);
	merge(data, l, m, r);
}

void merge(int* data, int l, int m, int r)
{
	const int leftSize = m - l + 1;
	const int rightSize = r - m;

	int* leftArr = new int[leftSize];
	int* rightArr = new int[rightSize];

	for (size_t i = 0; i < leftSize; i++)
	{
		leftArr[i] = data[l + i];
	}
	for (size_t i = 0; i < rightSize; i++)
	{
		rightArr[i] = data[m + i + 1];
	}

	int lInd = 0, rInd = 0, dInd = l;

	while (lInd < leftSize && rInd < rightSize)
	{
		if (leftArr[lInd] <= rightArr[rInd])
		{
			data[dInd] = leftArr[lInd];
			lInd++;
		}
		else
		{
			data[dInd] = rightArr[rInd];
			rInd++;
		}
		dInd++;
	}
	while (lInd < leftSize) {
		data[dInd] = leftArr[lInd];
		dInd++;
		lInd++;
	}

	delete[](leftArr);
	delete[](rightArr);
}

File: Java/MergeSort.java
import java.io.File;
import java.io.FileWriter;
import java.nio.file.Paths;
import java.util.ArrayList;
import java.util.List;
import java.util.Scanner;

class MergeSort {
  public static void main(String args[]) throws Exception {
    String input = args[0];
    String output = args[1];
    int iterations = Integer.parseInt(args[2]);
    List<Integer> list = new ArrayList<>();

    // Read from file
    File inFile = new File(input);
    Scanner scanner = new Scanner(inFile);
    while(scanner.hasNext()) {
      list.add(Integer.parseInt(scanner.nextLine().trim()));
    }
    scanner.close();

    // Sort and write to file
    File outFile = new File(output);
    outFile.createNewFile();
    FileWriter writer = new FileWriter(outFile);
    for(int i = 0; i < iterations; i++) {
      List<Integer> copy = new ArrayList<>(list);
      long start = System.nanoTime();
      sort(copy);
      long end = System.nanoTime();
      writer.write(Long.toString(end - start) + "\n");
    }
    writer.close();

  }

  private static void sort(List<Integer> list) {
    sort(list, 0, list.size() - 1);
  }

  private static void sort(List<Integer> list, int l, int r) {
    if (l >= r)
	  {
		  return;
	  }
    int m = l + (r - l) / 2;
    sort(list, l, m);
    sort(list, m + 1, r);
    merge(list, l, m, r);
  }

  private static void merge(List<Integer> list, int l, int m, int r) {
    int leftSize = m - l + 1;
	  int rightSize = r - m;
    
    Integer[] left = new Integer[leftSize];
    Integer[] right = new Integer[rightSize];

    for (int i = 0; i < leftSize; i++) {
      left[i] = list.get(i + l);
    }
    for (int i = 0; i < rightSize; i++) {
      right[i] = list.get(i + m + 1);
    }

    int lInd = 0;
    int rInd = 0;
    int dInd = l;

	  while (lInd < leftSize && rInd < rightSize)
	  {
		  if (left[lInd] <= right[rInd])
		  {
			  list.add(dInd, left[lInd]);
  			lInd++;
  		}
  		else
  		{
  			list.add(dInd, right[rInd]);
  			rInd++;
  		}
  		dInd++;
  	}
  	while (lInd < leftSize) {
  		list.add(dInd, left[lInd]);
  		dInd++;
  		lInd++;
  	}
  }
}
